lxd/device/gpu_sriov.go

```diff
@@ -0,0 +1,297 @@
+	"github.com/lxc/lxd/lxd/resources"
+	"github.com/lxc/lxd/lxd/revert"
+	"github.com/lxc/lxd/shared"
+)
+
+type gpuSRIOV struct {
+	deviceCommon
+}
+
+// validateConfig checks the supplied config for correctness.
+func (d *gpuSRIOV) validateConfig(instConf instance.ConfigReader) error {
+	if !instanceSupported(instConf.Type(), instancetype.VM) {
+		return ErrUnsupportedDevType
+	}
+
+	requiredFields := []string{}
+
+	optionalFields := []string{
+		"vendorid",
+		"productid",
+		"id",
+		"pci",
+	}
+	err := d.config.Validate(gpuValidationRules(requiredFields, optionalFields))
+	if err != nil {
+		return err
+	}
+
+	if d.config["pci"] != "" {
+		for _, field := range []string{"id", "productid", "vendorid"} {
+			if d.config[field] != "" {
+				return fmt.Errorf(`Cannot use %q when when "pci" is set`, field)
+			}
+		}
+	}
+
+	if d.config["id"] != "" {
+		for _, field := range []string{"pci", "productid", "vendorid"} {
+			if d.config[field] != "" {
+				return fmt.Errorf(`Cannot use %q when when "id" is set`, field)
+			}
+		}
+	}
+
+	return nil
+}
+
+// validateEnvironment checks the runtime environment for correctness.
+func (d *gpuSRIOV) validateEnvironment() error {
+	if d.config["pci"] != "" && !shared.PathExists(fmt.Sprintf("/sys/bus/pci/devices/%s", d.config["pci"])) {
+		return fmt.Errorf("Invalid PCI address (no device found): %s", d.config["pci"])
+	}
+
+	return nil
+}
+
+// Start is run when the device is added to the instance.
+func (d *gpuSRIOV) Start() (*deviceConfig.RunConfig, error) {
+	err := d.validateEnvironment()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get SRIOV parent, i.e. the actual GPU.
+	runConf := deviceConfig.RunConfig{}
+	gpus, err := resources.GetGPU()
+	if err != nil {
+		return nil, err
+	}
+
+	saveData := make(map[string]string)
+	var parentPCIAddress string
+
+	for _, gpu := range gpus.Cards {
+		// Skip any cards that don't match the vendorid, pci, productid or DRM ID settings (if specified).
+		if (d.config["vendorid"] != "" && gpu.VendorID != d.config["vendorid"]) ||
+			(d.config["pci"] != "" && gpu.PCIAddress != d.config["pci"]) ||
+			(d.config["productid"] != "" && gpu.ProductID != d.config["productid"]) ||
+			(d.config["id"] != "" && (gpu.DRM == nil || fmt.Sprintf("%d", gpu.DRM.ID) != d.config["id"])) {
+			continue
+		}
+
+		if parentPCIAddress != "" {
+			return nil, fmt.Errorf("VMs cannot match multiple GPUs per device")
+		}
+
+		parentPCIAddress = gpu.PCIAddress
+	}
+
+	if parentPCIAddress == "" {
+		return nil, fmt.Errorf("Failed to detect requested GPU device")
+	}
+
+	// Get PCI information about the GPU device.
+	devicePath := filepath.Join("/sys/bus/pci/devices", parentPCIAddress)
+
+	pciParentDev, err := pci.ParseUeventFile(filepath.Join(devicePath, "uevent"))
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get PCI device info for GPU %q", parentPCIAddress)
+	}
+
+	// Record parent PCI address.
+	saveData["last_state.vf.parent"] = pciParentDev.SlotName
```

Comment:
@monstermunchkin was there any reason why you didn't repeat the process here (https://github.com/lxc/lxd/pull/8375/files#diff-80e79853111c82d9d4ff94003e269ee7db2aa2f2ca4f2de8fd31d4ca994c3538R89-R103) to find the parent PCI address on device stop rather than storing it in volatile?

Reply:
@tomponline I wanted to avoid repetition. Alternatively, I could move the `for` loop to a separate function.

Reply:
@monstermunchkin yeah unless that process is particularly expensive, I would tend to prefer repeating the process on stop rather than storing more volatile keys. If you're primarily doing it to avoid code repetition then a separate function sounds good.

Reply:
@tomponline already done